tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Graph.java

```diff
@@ -138,14 +217,101 @@ public GraphOperation operation(String name) {
     return new OperationIterator(this);
   }
 
+  private GraphOperation graphOp(Operand<?> operand) {
+    checkInput(operand);
+    return (GraphOperation) operand.op();
+  }
+
+  /**
+   * Finds the operations used to produce {@code outputs} from {@code inputs}, or throws if that is not possible.
+   * Respects control dependencies.
+   *
+   * If both {@code allowedBodyOps} and {@code forbiddenBodyOps} are null, forbids 0-input ops in the body.
+   *
+   * @param inputs the inputs of the subgraph.  Must be from single output ops.
+   * @param outputs the outputs of the subgraph
+   * @param allowedBodyOps types of ops to allow as 0-input ops in the body.  Allows all (except {@code
+   * forbiddenBodyOps}) if null.
+   * @param forbiddenBodyOps types of ops to never allow as 0-input ops in the body.  Forbids all (except {@code
+   * allowedBodyOps}) if null.
+   * @return the set of operations needed to calculate outputs from inputs, including outputs and inputs
+   * @throws IllegalStateException if outputs depends on ops outside of the subgraph (i.e. is not calculable based
+   * solely on inputs)
+   */
+  public synchronized Set<GraphOperation> completeSubgraph(Set<Operand<?>> inputs, Set<Operand<?>> outputs,
+      Set<String> allowedBodyOps, Set<String> forbiddenBodyOps) {
+
+    if (forbiddenBodyOps != null && allowedBodyOps != null) {
+      for (String t : forbiddenBodyOps) {
+        if (allowedBodyOps.contains(t)) {
+          throw new IllegalArgumentException("Can't allow and forbid op type " + t + ".");
+        }
+      }
+    }
+
+    Queue<GraphOperation> currents = new LinkedList<>();
+    Set<GraphOperation> seen = new LinkedHashSet<>(outputs.size());
+    Set<GraphOperation> inputOps = new LinkedHashSet<>(inputs.size());
+
+    for (Operand<?> input : inputs) {
+      if (input.op().numOutputs() > 1) {
+        throw new IllegalStateException("Only ops with one output are supported as subgraph inputs");
+      }
+      GraphOperation op = graphOp(input);
+      inputOps.add(op);
+      seen.add(op);
+    }
+
+    for (Operand<?> operand : outputs) {
+      GraphOperation op = graphOp(operand);
+      currents.add(op);
+    }
+
+    while (!currents.isEmpty()) {
+      GraphOperation op = currents.poll();
+
+      // skip if already present
+      if (!seen.add(op)) {
+        continue;
+      }
+
+      if (op.numControlInputs() + op.numInputs() == 0) {
+        if ((forbiddenBodyOps != null && forbiddenBodyOps.contains(op.type()))
```

Comment:
Initialising these to `Collections.emptySet()` internally even if we don't make users do it will make this if a lot less ugly.

Reply:
See one of the other comments, nulls are necessary to distinguish from "only allow the empty set" which would error whenever something is in the body.  Optionally we could do that behavior (i.e. allow everything) for the empty set, but imo nulls is clearer (and what if that's what you want?).

Reply:
Why would we want the behaviour of only allowing the empty set? So you could catch the runtime exception and have different behaviour if there were intermediate nodes?

Anything that implies users have to catch a subclass of `RuntimeException` isn't good.

Reply:
It's for 0 input intermediate nodes specifically, i.e. things that don't depend on the inputs.  Generally `Constant` is ok and the others (`Placeholder`, `Variable`, etc) are ok, but since I'm making it a parameter "none" is an option.

Reply:
Ok. I'm still not sure that the semantics of this is quite right, but given we've not stabilised the API yet lets add it in and use it a bit first before making further decisions.